File: Leetcode/003. Longest Substring Without Repeating Characters/LongestSubstringWithoutRepeatingCharacters.java
// Leetcode 3. Longest Substring Without Repeating Characters
// https://leetcode.com/problems/longest-substring-without-repeating-characters/
// Runtime: 32 ms

class Solution 
{
public:
    int lengthOfLongestSubstring(string s) 
    {
        int sz = s.size();
        if (sz < 1)
        {
            return sz;
        }
        
        bool *tbl = (bool*)malloc(sizeof(bool) * 128);
        for (int i = 0; i < 128; i++)
        {
            tbl[i] = false;
        }
        
        int i = 0;
        int j = 1;
        
        int max = 1;
        tbl[s[0]] = true;
        while (i < sz && j < sz)
        {
            if (tbl[s[j]])
            {
                // found same
                for (int i = 0; i < 128; i++)
                {
                    tbl[i] = false;
                }
                
                if (j - i > max)
                {
                    max = j - i;
                }
                
                i++;
                tbl[s[i]] = true;
                j = i + 1;
            }
            else
            {
                tbl[s[j]] = true;
                j++;
            }
        }
        
        // final check
        if (j - i > max)
        {
            max = j - i;
        }
            
        return max;
    }
};

File: Leetcode/003. Longest Substring Without Repeating Characters/longest_substring_without_repeating_characters.cpp
// Leetcode 3. Longest Substring Without Repeating Characters
// https://leetcode.com/problems/longest-substring-without-repeating-characters/
// Runtime: 20 ms
// Memory: 7.6 MB

class Solution 
{
public:
    int lengthOfLongestSubstring(string s) 
    {
        int *tbl = (int*)malloc(sizeof(int) * 128);  // table for saving index
        for (int i = 0; i < 128; i++)
        {
            tbl[i] = -1;
        }
        
        int l = 0;
        int r = 0;
        int max = 0;

        while (r < s.size())
        {
            if (tbl[s[r]] != -1)
            {
                int len_l = tbl[s[r]] - l + 1;
                int len_r = r - tbl[s[r]];
                int tmp_max = len_l > len_r ? len_l : len_r;
                
                int allrange = r - l;  // case like "abcda", need to check the whole range
                tmp_max = tmp_max > allrange ? tmp_max : allrange;
                
                if (tmp_max > max)
                {
                    max = tmp_max;
                }
                
                // clear the table before repeated character
                int end = tbl[s[r]];
                for (int i = l; i < end; i++)
                {
                    tbl[s[i]] = -1;
                }
                
                // move left index next to the repeated character directly
                l = tbl[s[r]] + 1;
                tbl[s[r]] = r;
                r++;
            }
            else
            {
                // save the index of current character
                tbl[s[r]] = r;
                r++;
            }
        }
        
        // final check (missed this often)
        if (r - l > max)
        {
            max = r - l;
        }
        
        return max;
    }
};
